dmoj/cptbox/ptdebug_x64.cpp

```diff
@@ -2,96 +2,129 @@
+    if (ptrace(PTRACE_SETREGSET, tid, NT_PRSTATUS, &iovec))
+        perror("ptrace(PTRACE_SETREGSET)");
 }
 
-void pt_debugger_x64::result(long value) {
-    poke_reg(RAX, value);
+#define INVALID_ABI(source) fprintf(stderr, source ": Invalid ABI\n")
+
+int pt_debugger::syscall() {
+    switch (abi_) {
+        case PTBOX_ABI_X86:
+            return regs.x86.orig_eax;
+        case PTBOX_ABI_X32:
+            return regs.x64.orig_rax & ~__X32_SYSCALL_BIT;
+        case PTBOX_ABI_X64:
+            return regs.x64.orig_rax;
+        default:
+            INVALID_ABI("ptdebug_x64.cpp:syscall getter");
+            return -1;
+    }
 }
 
-#define make_arg(id, reg) \
-    long pt_debugger_x64::arg##id() { \
-        return peek_reg(reg); \
+void pt_debugger::syscall(int id) {
+    regs_changed = true;
+    switch (abi_) {
+        case PTBOX_ABI_X86:
+            regs.x86.orig_eax = id;
+            return;
+        case PTBOX_ABI_X32:
+            regs.x64.orig_rax = id | __X32_SYSCALL_BIT;
+            return;
+        case PTBOX_ABI_X64:
+            regs.x64.orig_rax = id;
+            return;
+        default:
+            INVALID_ABI("ptdebug_x64.cpp:syscall setter");
+    }
+}
+
+#define MAKE_ACCESSOR(method, x86_name, x64_name) \
+    long pt_debugger::method() { \
+        switch (abi_) { \
+            case PTBOX_ABI_X86: \
+                return regs.x86.x86_name; \
+            case PTBOX_ABI_X32: \
+            case PTBOX_ABI_X64: \
+                return regs.x64.x64_name; \
+            default: \
+                return INVALID_ABI("ptdebug_x64.cpp:" #method " getter"); \
+        } \
     } \
     \
-    void pt_debugger_x64::arg##id(long data) {\
-        poke_reg(reg, data); \
+    void pt_debugger::method(long value) { \
+        regs_changed = true; \
+        switch (abi_) { \
+            case PTBOX_ABI_X86: \
+                regs.x86.x86_name = value; \
+                return; \
+            case PTBOX_ABI_X32: \
+            case PTBOX_ABI_X64: \
+                regs.x64.x64_name = value; \
+                return; \
+            default: \
+                INVALID_ABI("ptdebug_x64.cpp:" #method " setter"); \
+        } \
     }
 
-make_arg(0, RDI);
-make_arg(1, RSI);
-make_arg(2, RDX);
-make_arg(3, R10);
-make_arg(4, R8);
-make_arg(5, R9);
-
-#undef make_arg
-
-bool pt_debugger_x64::is_exit(int syscall) {
-#if defined(__FreeBSD__)
-    return syscall == 1;
-#else
-    return syscall == 231 || syscall == 60;
-#endif
-}
+MAKE_ACCESSOR(result, eax, rax)
+MAKE_ACCESSOR(arg0, ebx, rdi)
+MAKE_ACCESSOR(arg1, ecx, rsi)
+MAKE_ACCESSOR(arg2, edx, rdx)
+MAKE_ACCESSOR(arg3, esi, r10)
+MAKE_ACCESSOR(arg4, edi, r8)
+MAKE_ACCESSOR(arg5, ebp, r9)
 
-int pt_debugger_x64::getpid_syscall() {
-#if defined(__FreeBSD__)
-    return 20;
-#else
-    return 39;
-#endif
-}
+#undef MAKE_ACCESSOR
 
-pt_debugger_x64::pt_debugger_x64() {
-    execve_id = 59;
+int pt_debugger::first_execve_syscall() {
```

Comment:
`execve_syscall_id` to match the naming of `(skip|noop)_syscall_id` from before?